x/wasm/internal/keeper/ioutil.go

```diff
@@ -0,0 +1,38 @@
+package keeper
+
+import (
+	"bytes"
+	"compress/gzip"
+	"io"
+	"io/ioutil"
+)
+
+// magic bytes to identify gzip.
+// See https://www.ietf.org/rfc/rfc1952.txt
+// and https://github.com/golang/go/blob/master/src/net/http/sniff.go#L186
+var gzipIdent = []byte("\x1F\x8B\x08")
+
+// limit max bytes read to prevent gzip bombs
+const maxSize = 400 * 1024
+
+// uncompress returns gzip uncompressed content or given src when not gzip.
+func uncompress(src []byte) ([]byte, error) {
+	if len(src) < 3 {
+		return src, nil
+	}
+	in := io.LimitReader(bytes.NewReader(src), maxSize)
+	buf := make([]byte, 3)
```

Comment:
This seems like a good way to ensure exactly 3 bytes read, but I wonder if it is guaranteed never to read an extra byte and drop it somewhere. https://golang.org/pkg/io/#ReadAtLeast

But I really doubt odd behavior when the input is a buffer not some network stream

Reply:
From the [Reader Godoc](https://golang.org/pkg/io/#Reader):
> Read reads up to len(p) bytes into p.

Reply:
Can you replace lines 21-30 with:

```go
if !bytes.Equal(gzipIdent, src[:3]) {
  return src, nil
}
zr, err := gzip.NewReader(bytes.NewReader(src))
```

Same functionality, just cleaner

Reply:
ja, much cleaner as src is not a stream. Thanks!